tests/test_aiohttp.py

```diff
@@ -0,0 +1,295 @@
+    query = str_query
+    resp = await client.get("/ws", params={"query": query})
+    assert resp.status == 200
+    result = await resp.json()
+    assert result == dict(data=dict(getChannel=dict(value=dict(string="longout"))))
+
+
+@pytest.mark.asyncio
+async def test_get_nan_pv(ioc: Popen, client, nan_query):
+    query = nan_query
+    val = await caget(PV_PREFIX + "nan")
+    assert math.isnan(val)
+    resp = await client.get("/ws", params={"query": query})
+    assert resp.status == 200
+    result = await resp.json()
+    assert result == dict(data=dict(getChannel=dict(value=dict(float=None))))
+
+
+@pytest.mark.asyncio
+async def test_get_enum_pv(ioc: Popen, client, enum_query):
+    query = enum_query
+    resp = await client.get("/ws", params={"query": query})
+    assert resp.status == 200
+    result = await resp.json()
+    assert result == dict(
+        data=dict(
+            getChannel=dict(
+                value=dict(string="nm", float=3.0),
+                display=dict(choices=["m", "mm", "um", "nm"]),
+            )
+        )
+    )
+
+
+@pytest.mark.asyncio
+async def test_put_long_and_enum(ioc: Popen, client, long_and_enum_put):
+    query = long_and_enum_put
+    resp = await client.post("/ws", json={"query": query})
+    assert resp.status == 200
+    result = await resp.json()
+    assert result == dict(
+        data=dict(
+            putChannels=[
+                dict(value=dict(string="55"), time=ANY),
+                dict(value=dict(string="mm"), time=ANY),
+            ]
+        )
+    )
+    thens = [
+        datetime.fromisoformat(r["time"]["datetime"])
+        for r in result["data"]["putChannels"]
+    ]
+    now = datetime.now()
+    for then in thens:
+        diff = now - then
+        # Shouldn't take more than this time to get the result of a put out
+        assert diff.total_seconds() < 0.2
+
+
+@pytest.mark.asyncio
+async def test_put_list(ioc: Popen, client, list_put):
+    query = list_put
+    resp = await client.post("/ws", json={"query": query})
+    assert resp.status == 200
+    result = await resp.json()
+    assert result == dict(
+        data=dict(
+            putChannels=[
+                dict(
+                    value=dict(
+                        stringArray=["0.0", "1.7", "2.0"], base64Array=BASE64_0_1688_2
+                    )
+                )
+            ]
+        )
+    )
+
+
+@pytest.mark.asyncio
+async def test_put_base64(ioc: Popen, client, base64_put):
+    # first dump gives {"key": "value"}, a json string
+    # second dump gives \{\"key\": \"value\"\}, an escaped json string
+    query = base64_put
+    resp = await client.post("/ws", json={"query": query})
+    assert resp.status == 200
+    result = await resp.json()
+    assert result == dict(
+        data=dict(putChannels=[dict(value=dict(stringArray=["0.0", "1.7", "2.0"]))])
+    )
+
+
+@pytest.mark.asyncio
+async def test_subscribe_ticking_graphql_transport_ws_protocol(
```

Comment:
I have been able to parametrize this test so that we simply have to define the difference between the websocket protocols as inputs to a generic function. See commit https://github.com/dls-controls/coniql/pull/44/commits/caf1cbd0a881cdd95001f7301a8b47c462a7f7e3.
Note however that this has been superseded in my attempts to parametrize both test_caplugin and test_aiohttp tests.